src/ir/module-splitting.cpp

```diff
@@ -0,0 +1,405 @@
+      secondaryItem.base = exportIt->second;
+    } else {
+      exportName = Names::getValidExportName(primary, exportName);
+      primary.addExport(new Export{exportName, primaryItem.name, kind});
+      secondaryItem.base = exportName;
+    }
+  };
+
+  // TODO: Be more selective by only sharing global items that are actually used
+  // in the secondary module, just like we do for functions.
+
+  auto shareMemory = [&](Memory* memory) {
+    secondary.memory.exists = true;
+    secondary.memory.initial = memory->initial;
+    secondary.memory.max = memory->max;
+    secondary.memory.shared = memory->shared;
+    secondary.memory.indexType = memory->indexType;
+    makeImportExport(*memory, secondary.memory, "memory", ExternalKind::Memory);
+  };
+  ModuleUtils::iterImportedMemories(primary, shareMemory);
+  ModuleUtils::iterDefinedMemories(primary, shareMemory);
+
+  auto shareTable = [&](Table* table) {
+    secondary.table.exists = true;
+    secondary.table.initial = table->initial;
+    secondary.table.max = table->max;
+    makeImportExport(*table, secondary.table, "table", ExternalKind::Table);
+  };
+  ModuleUtils::iterImportedTables(primary, shareTable);
+  ModuleUtils::iterDefinedTables(primary, shareTable);
+
+  auto shareGlobal = [&](Global* global) {
+    assert(primary.features.hasMutableGlobals() &&
+           "TODO: add wrapper functions for disallowed mutable globals");
+    auto secondaryGlobal = std::make_unique<Global>();
+    secondaryGlobal->type = global->type;
+    secondaryGlobal->mutable_ = global->mutable_;
+    secondaryGlobal->init =
+      global->init == nullptr
+        ? nullptr
+        : ExpressionManipulator::copy(global->init, secondary);
+    makeImportExport(*global, *secondaryGlobal, "global", ExternalKind::Global);
+    secondary.addGlobal(std::move(secondaryGlobal));
+  };
+  ModuleUtils::iterImportedGlobals(primary, shareGlobal);
+  ModuleUtils::iterDefinedGlobals(primary, shareGlobal);
+
+  auto shareEvent = [&](Event* event) {
+    auto secondaryEvent = std::make_unique<Event>();
+    secondaryEvent->attribute = event->attribute;
+    secondaryEvent->sig = event->sig;
+    makeImportExport(*event, *secondaryEvent, "event", ExternalKind::Event);
+    secondary.addEvent(std::move(secondaryEvent));
+  };
+  ModuleUtils::iterImportedEvents(primary, shareEvent);
+  ModuleUtils::iterDefinedEvents(primary, shareEvent);
+}
+
+template<class F> void forEachElement(Table& table, F f) {
+  for (auto& segment : table.segments) {
+    assert(segment.offset->is<Const>() &&
+           "TODO: handle non-constant segment offsets");
+    Index baseOffset = segment.offset->cast<Const>()->value.geti32();
+    for (size_t i = 0; i < segment.data.size(); ++i) {
+      f(Index(baseOffset + i), segment.data[i]);
+    }
+  }
+}
+
+// Return the first free table slot and the segment to which to append new
+// items, if it exists.
+Index getFirstFreeTableIndex(Table& table, Table::Segment** outSegment) {
```

Comment:
can't this use `forEachElement`?

Reply:
That would be overkill, since this doesn't need to access individual table slots, just the segments. The similarity is unfortunate, though.